### ISSUE TITLE
Where should AttributeCommand go?

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#112]
This module contains a tool which is a quite specific implementation of a basic-style-like command. We'll have more commands like that in the future. Should all of them end up in this repository?

It's not against any rule that some tools are keept in this repository, but we'll also face an issue of how much should be here. We'll need to consider whether such tools should not end up in a separate package(s). Otherwise this repository may grow fast what will also be problematic because of #111.

**Comment by scofalik** [comment]
Let me ask differently: what is the purpose of keeping `AttributeCommand` in main repository? What problem do we solve?

**Comment by pjasiun** [comment]
Can you name more than 3 basic commands? If not, I think we could keep it here. Whey it grow to much be should look for the part of the logic which is consistent and could be extracted from this repo.

**Comment by Reinmar** [comment]
I'm referring here to my following comments in the core:

> I wonder which pieces of this command could be useful outside it. Right now we assume that to operate on attributes you need this command, but what if your functionality needs to perform similar action but without registering a command?
>
> I see that either we can expose some of this command's functionality into some AttrbiuteApplier or we can tell – you want a piece of this command, then create it (what can be done without registering).

> It may be a bit too early to understand what's the best way to expose this logic, so we can postpone this decision and keep the current implementation.

> I'm looking at this class again and I feel that it will be crucial to make it (or related commands/classes) flexible and extensible. For instance, this command will need to share some logic with command which is able to set attribute with a specified value (e.g. the link command would need such a command). Then, we have base link feature extensions, like for instance link styles. I don't know yet how those extensions will be defined, either as separate commands or some logic pluggable into existing commands. Tough stuff to predict, so let's just keep this in our minds.

I think that these commands may grow, because so far they are encapsulating the logic of applying attributes, which will be reused in many ways. So it's actually a question of what we'll need to expose from that command and how much other code we'll need to add. We're looking here at reincarnation of CKEditor 4's `CKEDITOR.style` class which is a complex thing. We'll need similar mechanisms in CKE5 – flexible inline styles, object styles, widget styles, block styles, formats, etc. They cannot be exposed in a form of commands, because the code must be reusable without them.

I've opened this issue to remember about this topic, not to do any change right now. But I think that it will end like this:

1. Generic tools (replicating `CKEDITOR.style`) will be added to the engine. They will be exposed as simple methods/classes, not as commands, because they can't know about the editor state. They need to be focused on what the engine gives.
2. We'll have a couple of basic commands, probably in this repo. Since most of the logic of these commands will be kept in the engine, we can safely keep the commands here.

**Comment by Reinmar** [comment]
Turns out that I was right ;) During today's discussing with @pjasiun and @oskarwrobel we decided to extract generic code from `AttributeCommand` to some utils and simplify the command so it only handles toggling attribute. We can later add more command, but e.g. the [link feature](https://github.com/ckeditor/ckeditor5-link) will for now implement its own one, because we're unsure what can be reused there and whether it makes for a separate class or rather tools.
